src/main/java/picard/illumina/NewIlluminaBasecallsConverter.java

```diff
@@ -0,0 +1,393 @@
+    private final File[] filterFiles;
+    private final int maxNoCalls;
+    private final int maxMismatches;
+    private final int minMismatchDelta;
+    private final int minimumBaseQuality;
+    private final MetricsFile<BarcodeMetric, Integer> metrics;
+    private final File metricsFile;
+    private final Map<String, ThreadPoolExecutorWithExceptions> barcodeWriterThreads = new HashMap<>();
+    private final Map<Integer, List<RecordWriter>> completedWork = new HashMap<>();
+
+    /**
+     * @param basecallsDir             Where to read basecalls from.
+     * @param barcodesDir              Where to read barcodes from (optional; use basecallsDir if not specified).
+     * @param lane                     What lane to process.
+     * @param readStructure            How to interpret each cluster.
+     * @param barcodeRecordWriterMap   Map from barcode to CLUSTER_OUTPUT_RECORD writer.  If demultiplex is false, must contain
+     *                                 one writer stored with key=null.
+     * @param demultiplex              If true, output is split by barcode, otherwise all are written to the same output stream.
+     * @param maxReadsInRamPerTile     Configures number of reads each tile will store in RAM before spilling to disk.
+     * @param tmpDirs                  For SortingCollection spilling.
+     * @param numProcessors            Controls number of threads.  If <= 0, the number of threads allocated is
+     *                                 available cores - numProcessors.
+     * @param firstTile                (For debugging) If non-null, start processing at this tile.
+     * @param tileLimit                (For debugging) If non-null, process no more than this many tiles.
+     * @param outputRecordComparator   For sorting output records within a single tile.
+     * @param codecPrototype           For spilling output records to disk.
+     * @param outputRecordClass        Inconveniently needed to create SortingCollections.
+     * @param includeNonPfReads        If true, will include ALL reads (including those which do not have PF set)
+     * @param ignoreUnexpectedBarcodes If true, will ignore reads whose called barcode is not found in barcodeRecordWriterMap,
+     * @param maxNoCalls               The maximum number of no calls to allow in a match.
+     * @param maxMismatches            The maximum number of mismatched calls to allow in a match.
+     * @param minMismatchDelta         The minimum mismatch difference between the best and second best matches.
+     * @param minimumBaseQuality       The minimum base quality to allow for a matching base call.
+     * @param metrics                  The metrics output for barcode metrics.
+     * @param metricsFile              The metrics file to write out the barcode metrics to.
+     */
+    public NewIlluminaBasecallsConverter(final File basecallsDir, File barcodesDir, final int lane,
```

Comment:
'NewIlluminaBasecallsConverter?' - thought this was flagged previously as an unimaginative name.